File: lib/domain/backdoor/gateways/application_database_clear_gateway.rb
# frozen_string_literal: true

module Domain
  module Backdoor
    module Gateways
      # バックドア用途: 匿名ユーザー以外のアプリデータを一括削除（永続化は Adapter）。
      module ApplicationDatabaseClearGateway
        ApplicationDataStats = Struct.new(:users, :farms, :fields, :crops, :cultivation_plans, keyword_init: true)

        ClearResult = Struct.new(:kind, :before_stats, :after_stats, :error_message, keyword_init: true) do
          def self.success(before:, after:)
            new(kind: :success, before_stats: before, after_stats: after)
          end

          def self.failure(message)
            new(kind: :failure, error_message: message.to_s)
          end
        end

        def clear_application_data_preserving_anonymous_users
          raise NotImplementedError, "#{self.class} must implement clear_application_data_preserving_anonymous_users"
        end
      end
    end
  end
end

File: crates/agrr-domain/src/backdoor/gateways/application_database_clear_gateway.rs
/// Ruby: `ApplicationDatabaseClearGateway::ApplicationDataStats`
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ApplicationDataStats {
    pub users: i64,
    pub farms: i64,
    pub fields: i64,
    pub crops: i64,
    pub cultivation_plans: i64,
}

/// Ruby: `ApplicationDatabaseClearGateway::ClearResult`
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ClearApplicationDataResult {
    Success {
        before_stats: ApplicationDataStats,
        after_stats: ApplicationDataStats,
    },
    Failure {
        error_message: String,
    },
}

impl ClearApplicationDataResult {
    pub fn success(before: ApplicationDataStats, after: ApplicationDataStats) -> Self {
        Self::Success {
            before_stats: before,
            after_stats: after,
        }
    }

    pub fn failure(message: impl Into<String>) -> Self {
        Self::Failure {
            error_message: message.into(),
        }
    }
}

/// Ruby: `Domain::Backdoor::Gateways::ApplicationDatabaseClearGateway`
pub trait ApplicationDatabaseClearGateway: Send + Sync {
    fn clear_application_data_preserving_anonymous_users(&self) -> ClearApplicationDataResult;
}
